src/md_graph.cpp

```diff
@@ -476,215 +476,255 @@ static std::vector<DefaultKernelArg> WinogradNodeArgs()
+                                                            "padded_x === (x ~ 6)"};
+                add_meta_wino(map_wino_conv_s2_modd_xe3, 100);
+                g.AddEdge(nullptr, vc_s2, map_wino_conv_s2_modd_xe3);
+
+                add_relu(program_s2, kernel, vc_s2, WinogradNodeArgs, supported_arch);
+            }
+        }
 
-            FusionMDGraph_Edge_Map map_wino_conv_s2_modd;
-            map_wino_conv_s2_modd["constraints"] = {
-                "stride_h == 2", "padded_y === (y ~ 6)", "(x % 6) != 1", "padded_x === (x ~ 6)",
+        // Fused Winograd v21_1_2
+        {
+            auto add_meta_wino = [&](FusionMDGraph_Edge_Map& m, int weight) {
+                m["constraints"].emplace_back("weight === " + std::to_string(weight));
+                m["constraints"].emplace_back("oH * oW <= (2^23)");
+                m["constraints"].insert(
+                    m["constraints"].end(), common_constr.begin(), common_constr.end());
             };
 
-            add_meta_wino(map_wino_conv_s2_modd, 5);
-            g.AddEdge(nullptr, vc_s2, map_wino_conv_s2_modd);
-
-            // high priority edge for 3x3 kernels
-            FusionMDGraph_Edge_Map map_wino_conv_s2_modd_xe3;
-            map_wino_conv_s2_modd_xe3["constraints"] = {
-                "stride_h == 2",
-                "(x == 3) & (y == 3)",
-                "padded_y === (y ~ 6)",
-                "(x % 6) != 1",
-                "padded_x === (x ~ 6)",
+            auto add_v21_wino = [&](const std::string family,
+                                    const std::vector<std::string> supported_arch,
+                                    const int stride) {
+                const auto kernel_postfix = "_fp32_stride" + std::to_string(stride);
+                const auto kernel_file    = "Conv_Winograd_v21_1_2" + kernel_postfix + ".s";
+                const auto kernel_name    = "miopenSp3AsmConv_v21_1_2_" + family + kernel_postfix;
+
+                auto vc = std::make_shared<MDGraph_vertex>(
+                    miopenFusionOpConvForward, kernel_file, kernel_name, algo);
+                vc->solver         = solver::ConvBinWinogradRxSf2x3g1Fused{};
+                vc->default_args   = WinogradV21NodeArgs();
+                vc->supported_arch = supported_arch;
+
+                const auto stride_constr = "stride_h == " + std::to_string(stride);
+
+                FusionMDGraph_Edge_Map map_wino_conv;
+                map_wino_conv["constraints"] = {stride_constr};
+                add_meta_wino(map_wino_conv, 5);
+                g.AddEdge(nullptr, vc, map_wino_conv);
+
+                // add 3x3 with higher priority since its the fastest case
+                FusionMDGraph_Edge_Map map_wino_conv_xe3;
+                map_wino_conv_xe3["constraints"] = {stride_constr, "(y == 3) & (x == 3)"};
+                add_meta_wino(map_wino_conv_xe3, 100);
+                g.AddEdge(nullptr, vc, map_wino_conv_xe3);
```

Comment:
FYI winograd might not be the best for some configurations even if the filter size is 3x3

Reply:
You mean v21 Winograd could be slower than older fused Winograd?